core/pkg/artifacts/saver.go

```diff
@@ -467,10 +466,12 @@ func (as *ArtifactSaver) uploadMultipart(
 			"Content-Length:" + strconv.FormatInt(task.Size, 10),
 			"Content-Type:" + contentType,
 		}
-		task.SetCompletionCallback(func(t *filetransfer.Task) {
-			partResponses <- partResponse{partNumber: partData[i].PartNumber, task: t}
-			wg.Done()
-		})
+		task.TaskCompletionCallback = filetransfer.TaskCompletionCallback{
```

Comment:
I believe you should be able to just do

```go
task.TaskCompletionCallback.CompletionCallback = func() {
	...
}
```

That's still a bit verbose. Have you tried defining `TaskCompletionCallback` as a newtype instead?

```go
type TaskCompletionCallback func(fts FileTransferStats)

type SomeTask struct {
	OnCompleted TaskCompletionCallback
	...
}

func (t TaskCompletionCallback) Complete(fts FileTransferStats) {
	if t == nil {
		return
	}
	...
}
```

Then you should be able to write

```go
task.OnCompleted = func() {
	...
}
```

Reply:
ah actually I also realized that in these callback functions, the task we use is local. so for example when I push `task.Err` to a channel in a completion callback it'll actually just be the local `task.Err`, which will always be `nil`

I changed it to take in a task like it did originally, like this instead for example:
```golang
type DefaultTask struct {
    OnComplete func(*DefaultTask)
    ...
}

type DefaultUploadTask DefaultTask

func (t *DefaultUploadTask) Complete(fts FileTransferStats) {
    if t.OnComplete != nil {
        t.OnComplete((*DefaultTask)(t))
    }
    ...
}
```

let me know if you have any concerns/suggestions with this method!

Reply:
> the task we use is local

I might be missing something.. it's the same task object isn't it? It's passed by reference, not copied?

Reply:
oh my bad, I looked into this more and you're right- I thought it would be copied not passed by reference. won't make that change then